pkg/receive/handler.go

```diff
@@ -933,25 +916,138 @@ func (a expectedErrors) Len() int           { return len(a) }
 func (a expectedErrors) Less(i, j int) bool { return a[i].count < a[j].count }
 
-// determineWriteErrorCause extracts a sentinel error that has occurred more than the given threshold from a given fan-out error.
-// It will inspect the error's cause if the error is a MultiError,
-// It will return cause of each contained error but will not traverse any deeper.
-func determineWriteErrorCause(err error, threshold int) error {
+// errorSet is a set of errors.
+type errorSet struct {
+	reasonSet map[string]struct{}
+	errs      []error
+}
+
+// Error returns a string containing a deduplicated set of reasons.
+func (es errorSet) Error() string {
+	if len(es.reasonSet) == 0 {
+		return ""
+	}
+	reasons := make([]string, 0, len(es.reasonSet))
+	for reason := range es.reasonSet {
+		reasons = append(reasons, reason)
+	}
+	sort.Strings(reasons)
+
+	var buf bytes.Buffer
+	if len(reasons) > 1 {
+		fmt.Fprintf(&buf, "%d errors: ", len(es.reasonSet))
+	}
+
+	var more bool
+	for _, reason := range reasons {
+		if more {
+			buf.WriteString("; ")
+		}
+		buf.WriteString(reason)
+		more = true
+	}
+
+	return buf.String()
+}
+
+// Add adds an error to the errorSet.
+func (es *errorSet) Add(err error) {
 	if err == nil {
-		return nil
+		return
 	}
 
-	unwrappedErr := errors.Cause(err)
-	errs, ok := unwrappedErr.(errutil.NonNilMultiError)
-	if !ok {
-		errs = []error{unwrappedErr}
+	if len(es.errs) == 0 {
+		es.errs = []error{err}
+	} else {
+		es.errs = append(es.errs, err)
+	}
+	if es.reasonSet == nil {
+		es.reasonSet = make(map[string]struct{})
 	}
-	if len(errs) == 0 {
+
+	switch werr := err.(type) {
+	case *replicationErrors:
+		for reason := range werr.reasonSet {
+			es.reasonSet[reason] = struct{}{}
+		}
+	case *writeErrors:
+		for reason := range werr.reasonSet {
+			es.reasonSet[reason] = struct{}{}
+		}
+	default:
+		es.reasonSet[err.Error()] = struct{}{}
+	}
+}
+
+// writeErrors contains all errors that have
+// occurred during a remote-write request.
+type writeErrors struct {
+	errorSet
+}
+
+// ErrOrNil returns the writeErrors instance if any
+// errors are contained in it.
+// Otherwise, it returns nil.
+func (es *writeErrors) ErrOrNil() error {
+	if len(es.errs) == 0 {
 		return nil
 	}
+	return es
+}
 
-	if threshold < 1 {
-		return err
+// Cause returns the primary cause for a write failure.
+// If multiple errors have occurred, Cause will prefer
+// recoverable over non-recoverable errors.
+func (es *writeErrors) Cause() error {
+	if len(es.errs) == 0 {
+		return nil
+	}
+
+	expErrs := expectedErrors{
+		{err: errUnavailable, cause: isUnavailable},
+		{err: errNotReady, cause: isNotReady},
+		{err: errConflict, cause: isConflict},
+	}
+
+	var recoverableErr error
+	for _, werr := range es.errs {
+		var found bool
+		cause := errors.Cause(werr)
+		for _, exp := range expErrs {
+			if exp.cause(cause) {
+				found = true
+				exp.count++
+			}
+		}
+		if !found {
+			recoverableErr = cause
```

Comment:
If we return `internalErr` we would lose the original cause, but I guess that's fine because we can only handle 3 specific errors and everything else is treated as a generic internal error. Keep in mind that we always keep all reasons so we would send all error messages in the response. The cause method merely decides which error code to send back.